toolchain/check/subst.cpp

```diff
@@ -256,9 +261,10 @@ static auto Rebuild(Context& context, Worklist& worklist, SemIR::InstId inst_id,
   // Note that we pop in reverse order because we pushed them in forwards order.
   int32_t arg1 = PopOperand(context, worklist, inst.arg1_and_kind());
   int32_t arg0 = PopOperand(context, worklist, inst.arg0_and_kind());
-  auto type_id = inst.type_id().has_value()
-                     ? context.types().GetTypeIdForTypeInstId(worklist.Pop())
-                     : SemIR::TypeId::None;
+  auto type_id =
+      inst.type_id().has_value()
+          ? callbacks.RebuildType(SemIR::TypeInstId::UnsafeMake(worklist.Pop()))
```

Comment:
I think it would be best to avoid the UnsafeMake here. We believe it should be a type but there's nothing here that guarantees that locally by construction, or by checks. And in the presence of bugs that we write in the future, going through an unsafe incorrectly makes a buggy situation into a much more confusing buggy situation.
```suggestion
          ? callbacks.RebuildType(context.types().GetAsTypeInstId(worklist.Pop()))
```